src/promptflow/promptflow/_sdk/_utils.py

```diff
@@ -585,20 +584,43 @@ def _generate_tool_meta(
     timeout: int,
     *,
     include_errors_in_output: bool = False,
+    load_in_subprocess: bool = True,
 ) -> Dict[str, dict]:
+    """Generate tool meta from files.
+
+    :param flow_directory: flow directory
+    :param tools: tool list
+    :param raise_error: whether raise error when generate meta failed
+    :param timeout: timeout for generate meta
+    :param include_errors_in_output: whether include errors in output
+    :param load_in_subprocess: whether load tool meta with subprocess to prevent system path disturb. Default is True.
+        If set to False, will load tool meta in sync mode and timeout need to be handled outside current process.
+    :return: tool meta dict
+    """
     logger = logging.getLogger(LOGGER_NAME)
-    # use multi process generate to avoid system path disturb
-    manager = multiprocessing.Manager()
-    tools_dict = manager.dict()
-    exception_dict = manager.dict()
-    p = multiprocessing.Process(
-        target=_generate_meta_from_files, args=(tools, flow_directory, tools_dict, exception_dict)
-    )
-    p.start()
-    p.join(timeout=timeout)
-    if p.is_alive():
-        p.terminate()
-        p.join()
+    if load_in_subprocess:
+        # use multiprocess generate to avoid system path disturb
+        manager = multiprocessing.Manager()
+        tools_dict = manager.dict()
+        exception_dict = manager.dict()
+        p = multiprocessing.Process(
+            target=_generate_meta_from_files, args=(tools, flow_directory, tools_dict, exception_dict)
+        )
+        p.start()
+        p.join(timeout=timeout)
+        if p.is_alive():
+            logger.warning(f"Generate meta timeout after {timeout} seconds, terminate the process.")
+            p.terminate()
+            p.join()
+    else:
+        tools_dict, exception_dict = {}, {}
+
+        #  There is no built-in method to forcefully stop a running thread/coroutine in Python
+        #  because abruptly stopping a thread can cause issues like resource leaks,
+        #  deadlocks, or inconsistent states.
+        #  Caller needs to handle the timeout outside current process.
+        _generate_meta_from_files(tools, flow_directory, tools_dict, exception_dict)
```

Comment:
print warning if specify timeout?

Reply:
Haven't done so as not sure if it will pollute the stdout of pf util. Extension now depends on stdout to communicate with pf util to my understanding

Reply:
will confirm with related people

Reply:
I raise comment, because in such code branch, timeout is ignored.

Reply:
confirmed with extension that pfutil will dump return value to file now, warning added.